packages/codemirror/src/codemirror.js

```diff
@@ -300,43 +303,69 @@ export function createCypherEditor(parentDOMElement, options = {}) {
     fire: fireKeyDown
   } = createEventHandlers();
 
+  const { on: onKeyUp, off: offKeyUp, fire: fireKeyUp } = createEventHandlers();
+
   const lineNumberClick = (line, event) => {
     fireLineNumberClick(line, event);
   };
 
   const keyDown = (event) => {
+    pressedKey =
+      !event.metaKey &&
+      !event.ctrlKey &&
+      !event.altKey &&
+      event.key &&
+      event.key.length === 1
+        ? event.key
+        : null;
     fireKeyDown(event);
   };
 
-  const handleValueChanged = (value, changes) => {
+  const keyUp = (event) => {
+    if (deferredAutocomplete) {
+      showAutocomplete();
+      deferredAutocomplete = false;
+    }
+
+    pressedKey = null;
+    fireKeyUp(event);
+  };
+
+  const deferAutocompleteForChanges = (changes) => {
+    const { length = 0, newLength = 0 } = changes;
+    const isSmallInsert = newLength > length && newLength - length <= 2;
     if (
+      pressedKey &&
       cypherLanguage &&
       autocomplete &&
-      Array.isArray(autocompleteTriggerStrings)
+      Array.isArray(autocompleteTriggerStrings) &&
+      isSmallInsert
     ) {
-      let changedText = [];
+      let changedText = null;
       changes.iterChanges((fromA, toA, fromB, toB, inserted) => {
```

Comment:
I've not yet fully read up on how this changes object works/contains, but I'm assuming this code looks for the last bit of changed text on a single line (?)

Reply:
the codemirror api is complicated but mostly documented. but basically this code is trying to just get access to what changed as a single string. We only care about the first change, but the codemirror api seems to require iterating to get it...